yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +102,180 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
+        delegationTokenRenewerThread, timeToRenewal, TimeUnit.MILLISECONDS)
+    }
+  }
+
+  override def updateCredentialsIfRequired(): Unit = {
+    try {
+      val credentialsFile = sparkConf.get("spark.yarn.credentials.file")
+      if (credentialsFile != null && !credentialsFile.isEmpty) {
+        val remoteFs = FileSystem.get(conf)
+        val sparkStagingDir = System.getenv("SPARK_YARN_STAGING_DIR")
+        val stagingDirPath = new Path(remoteFs.getHomeDirectory, sparkStagingDir)
+        val credentialsFilePath = new Path(stagingDirPath, credentialsFile)
+        // If an update is currently in progress, come back later!
+        if (remoteFs.exists( new Path(stagingDirPath, "_UPDATING"))) {
+          delegationTokenRenewer.schedule(delegationTokenExecuterUpdaterThread, 1, TimeUnit.HOURS)
+        }
+        // Now check if the file exists, if it does go get the credentials from there
+        if (remoteFs.exists(credentialsFilePath)) {
+          val status = remoteFs.getFileStatus(credentialsFilePath)
+          val modTimeAtStart = status.getModificationTime
+          if (modTimeAtStart > lastCredentialsRefresh) {
+            val newCredentials = getCredentialsFromHDFSFile(remoteFs, credentialsFilePath)
+            val newStatus = remoteFs.getFileStatus(credentialsFilePath)
+            // File was updated after we started reading it, lets come back later and try to read
+            // it.
+            if (newStatus.getModificationTime != modTimeAtStart) {
+              delegationTokenRenewer
+                .schedule(delegationTokenExecuterUpdaterThread, 1, TimeUnit.HOURS)
+            } else {
+              UserGroupInformation.getCurrentUser.addCredentials(newCredentials)
+              lastCredentialsRefresh = status.getModificationTime
+              val totalValidity = getLatestValidity - lastCredentialsRefresh
+              val timeToRunRenewal = lastCredentialsRefresh + (0.8 * totalValidity).toLong
+              val timeFromNowToRenewal = timeToRunRenewal - System.currentTimeMillis()
+              delegationTokenRenewer.schedule(delegationTokenExecuterUpdaterThread,
+                timeFromNowToRenewal, TimeUnit.MILLISECONDS)
+            }
+          } else {
+            // Check every hour to see if new credentials arrived.
+            delegationTokenRenewer.schedule(delegationTokenExecuterUpdaterThread, 1, TimeUnit.HOURS)
+          }
+        }
+      }
+    } catch {
+      // Since the file may get deleted while we are reading it, catch the Exception and come
+      // back in an hour to try again
+      case e: Exception =>
+        logWarning(
+          "Error encountered while trying to update credentials, will try again in 1 hour", e)
+        delegationTokenRenewer.schedule(delegationTokenExecuterUpdaterThread, 1, TimeUnit.HOURS)
+    }
+  }
+
+  private[spark] def getCredentialsFromHDFSFile(
+    remoteFs: FileSystem,
+    tokenPath: Path
+  ): Credentials = {
```

Comment:
nit: I think the style is for these to go on the previous line (and for the previous lines to be double-indented). That is, if all this doesn't fit in a single line...